src/anemoi/graphs/generate/icon_mesh.py

```diff
@@ -0,0 +1,553 @@
+        else:
+            self.max_level = reflvl_cell.max()
+
+        # restrict to level `max_level`:
+        self.select_c = np.argwhere(reflvl_cell <= self.max_level)
+        # generate source grid node set:
+        self.nodeset = NodeSet(clon[self.select_c], clat[self.select_c])
+
+        if multi_mesh is not None:
+            # generate edges between source grid nodes and multi-mesh nodes:
+            edge_vertices = self._get_grid2mesh_edges(self.select_c, multi_mesh=multi_mesh)
+            super().__init__((self.nodeset, multi_mesh.nodeset), edge_vertices)
+
+    def _get_grid2mesh_edges(self, select_c: np.ndarray, multi_mesh: ICONMultiMesh) -> np.ndarray:
+        """Create "grid-to-mesh" edges, ie. edges from (clat,clon) to the
+        vertices of the multi-mesh.
+        """
+
+        num_cells = select_c.shape[0]
+        num_verts_per_cell = multi_mesh.cell_vertices.shape[1]
+        src_list = np.kron(np.arange(num_cells), np.ones((1, num_verts_per_cell), dtype=np.int64)).flatten()
+        dst_list = multi_mesh.cell_vertices[select_c[:, 0], :].flatten()
+        edge_vertices = np.stack((src_list, dst_list), axis=1, dtype=np.int64)
+        return edge_vertices
+
+
+# -------------------------------------------------------------
+
+
+@typechecked
+def get_icon_mesh_and_grid(
+    iverbosity: int, grid_file: str, max_level_multimesh: int, max_level_dataset: int
+) -> tuple[ICONMultiMesh, ICONCellDataGrid]:
+    """Factory function, creating an ICON multi-mesh and an ICON cell-grid."""
+    return (
+        multi_mesh := ICONMultiMesh(grid_file, max_level=max_level_multimesh, iverbosity=iverbosity),
+        ICONCellDataGrid(grid_file, multi_mesh, max_level=max_level_dataset, iverbosity=iverbosity),
+    )
+
+
+@typechecked
+def get_edge_attributes(
+    gc_send: np.ndarray, cc_send: np.ndarray, gc_recv: np.ndarray, cc_recv: np.ndarray
+) -> np.ndarray:
+    """Build edge features using the position on the unit sphere of the mesh
+    nodes.
+
+    The features are (4 input features in total):
+
+    - length of the edge
+    - vector difference between the 3d positions of the sender node and the
+      receiver node computed in a local coordinate system of the receiver.
+
+    The local coordinate system of the receiver is computed by applying two elemental
+    rotations:
+
+    - the first rotation changes the azimuthal angle until that receiver
+      node lies at longitude 0,
+    - the second rotation changes the polar angle until that receiver
+      node lies at latitude 0.
+
+    Literature: Appdx. A2.4 in
+       Lam, R. et al. (2022). GraphCast: Learning skillful medium-range
+       global weather forecasting. 10.48550/arXiv.2212.12794.
+
+    """
+
+    phi_theta = gc_recv  # precession and nutation angle
+    theta__phi = np.fliplr(phi_theta)
+    theta__phi[:, 1] *= -1.0  # angles = [theta, -phi]
+    # rotation matrix:
+    R = scipy.spatial.transform.Rotation.from_euler(seq="YZ", angles=theta__phi)
```

Comment:
nit: not a fan of single-letter capital variable names

Reply:
This has been handled by the commit 6ad883f073da15d9a70ac80d2c670dfeb454523e.